src/adapter/http/src/data/tail_handler.rs

```diff
@@ -56,60 +65,71 @@ pub async fn dataset_tail_handler(
             QueryError::Internal(e) => e.api_err(),
         })?;
 
-    let schema = if params.include_schema {
-        Some(
-            super::query_types::serialize_schema(df.schema().as_arrow(), params.schema_format)
-                .api_err()?,
-        )
-    } else {
-        None
-    };
+    let schema = params
+        .schema_format
+        .map(|fmt| Schema::new(df.schema().inner().clone(), fmt));
 
     let record_batches = df.collect().await.int_err().api_err()?;
     let json = super::query_types::serialize_data(&record_batches, params.data_format).api_err()?;
     let data = serde_json::value::RawValue::from_string(json).unwrap();
 
-    Ok(Json(TailResponseBody { data, schema }))
+    Ok(Json(Response {
+        data,
+        data_format: params.data_format,
+        schema,
+        schema_format: params.schema_format,
+    }))
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 // TODO: Sanity limits
-#[derive(Debug, serde::Deserialize)]
+#[derive(Debug, serde::Deserialize, utoipa::IntoParams)]
 #[serde(rename_all = "camelCase")]
-pub struct TailRequestParams {
+pub struct RequestParams {
+    /// Nuber of leading records to skip when returning result (used for
+    /// pagination)
     #[serde(default)]
-    skip: u64,
-    #[serde(default = "TailRequestParams::default_limit")]
-    limit: u64,
-    #[serde(alias = "format")]
-    #[serde(default)]
-    data_format: DataFormat,
+    pub skip: u64,
+
+    /// Maximum nuber of records to return (used for pagination)
+    #[serde(default = "RequestParams::default_limit")]
+    pub limit: u64,
+
+    /// How the output data should be ecoded
     #[serde(default)]
-    schema_format: SchemaFormat,
-    #[serde(alias = "schema")]
-    #[serde(default = "TailRequestParams::default_include_schema")]
-    include_schema: bool,
+    pub data_format: DataFormat,
+
+    /// How to encode the schema of the result
+    pub schema_format: Option<SchemaFormat>,
 }
 
-impl TailRequestParams {
+impl RequestParams {
     fn default_limit() -> u64 {
         100
     }
-
-    fn default_include_schema() -> bool {
-        true
-    }
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-#[derive(Debug, serde::Serialize)]
+#[derive(Debug, serde::Serialize, utoipa::ToSchema)]
+#[schema(as = DatasetTailResponse)]
 #[serde(rename_all = "camelCase")]
-pub struct TailResponseBody {
-    data: Box<serde_json::value::RawValue>,
+pub struct Response {
+    /// Resulting data
+    #[schema(value_type = Object)]
+    pub data: Box<serde_json::value::RawValue>,
+
+    /// How data is layed out in the response
+    pub data_format: DataFormat,
+
+    /// Schema of the resulting data
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub schema: Option<Schema>,
```

Comment:
Normally this is a default in Schema-first flows around Swagger. Are you sure it's necessary?

Reply:
Not sure what you are referring to here. `skip_serializing_if` is a `serde` instruction and `serde` will serialize `null` by default for empty option.